### DISCUSSION TITLE
How do you debug Rust code on Android?

### DISCUSSION THREAD

**Discussion by coder0xff** [fzyzcjy/flutter_rust_bridge#932] [Q&A]
Hi, I'm using flutter_rust_bridge to work on an application that I'm primarily testing on Android, and using Visual Studio Code for debugging the Flutter parts. I'm wondering if there's a way to debug the Rust parts invoked by Flutter in the same dev environment. Barring that, is there a way to debug the rust code in any of the other environments, e.g. Linux or Windows, Android Studio or GDB? Thanks for any advice.

**Comment by fzyzcjy** [comment]
Looking forward others to answer this, since I myself do not use debug (but use logging) when I want to see bugs in my code.

**Reply by coder0xff** [reply]
I tried using println! but nothing appeared, so I also tried the log+logsys crates to log to the system log on Android, but that was crashing due to the location not existing in the filesystem. Do you have a trick for Android log statements?

**Reply by fzyzcjy** [reply]
https://cjycode.com/flutter_rust_bridge/feature/logging.html

**Reply by coder0xff** [reply]
Thanks, I'll give it a try.

**Comment by tmpfs** [comment]
This worked for me:

```yaml
[target.'cfg(target_os="android")'.dependencies]
android_logger = "0.11.1"
```

Along with the usual `log` crate and then just `log::info!()` etc. Now I just need to figure out how to get rid of all the noise the android logs by default!

**Comment by coder0xff** [comment]
Not Android, but I found that for the Linux target that (after running `sudo sysctl kernel.yama.ptrace_scope=0`) I could go to the VS Code command menu with Ctrl+P > `LLDB: Attach to Process...` and attach to the process, giving me working breakpoints in Dart and Rust simultaneously.

Android Studio also has the ability to attach to running processes on a device, so I opened the android folder and gave it a try. It was able to attach to the JVM debugger, but when I tried to force native debugging it would show

![image](https://user-images.githubusercontent.com/2261949/216858782-016c95a1-926a-4da3-ad00-7f07e6ccf5d8.png)

If I were to hazard a guess, it doesn't know how to find the Rust code.
